### ISSUE TITLE
Critical Security Issue: "Error138" Backdoor

### ISSUE THREAD

**Issue by kaebasi** [WhiteHatSecurity/Aviator#24]
**URGENT!!!!!!!!!!!!** exploit on from twitter there

https://github.com/WhiteHatSecurity/Aviator/blob/master/src/chrome/browser/championconfig/ConnectionControl/ConnectionControlManager.cc#L166

osx: https://www.google.com/;/Applications/Calculator.app/Contents/MacOS/Calculator;:Error138
win: http://preview.tinyurl.com/ml69fsy

`ಠ_ಠ`

**Comment by ajinabraham** [comment]
Whatever its just hilarious.. The most secure browser, just got screwed by tarvis.

**Comment by schnauss** [comment]
"Yup, Patches welcome, it’s open source"

Utterly embarrassing.

**Comment by DinisCruz** [comment]
I think it is a great move and it really shows that when software is developed in a close source environment, this kind of bug is easily created, code-reviewed and released.

At least now one can do something about it

**Comment by cure53** [comment]
Well, here is another interpretation of how Aviator made it here, to be an open source tool. 

Or at least a hypothesis with partial proof and some educated guesses.

It's not meant to be an attack or an attempt to hurt anyone. I just want to understand how it came to this point and therefore share my thoughts and research results to be debated or disproved. 

Oh, and admittedly my assumptions are opinionated as I am a bit upset right now. 

 * About 1.5 years ago, some folks at WhiteHat Security came up with the idea of building/branding a browser
 * Codename for that project is "Champion". (see Aviator sources)
 * Focus of that project would be privacy and security as this matches WhiteHat's other product lines.
 * The future business model behind it would be a paid premium version of the browser. [Source](https://blog.whitehatsec.com/whitehat-aviator-beta-for-windows/)
 * WhiteHat get Robert Hansen on board for he is known to have a good background in web- and browser-security [Source](https://www.linkedin.com/profile/view?id=1000399)
 * The idea gets fleshed out and WhiteHat Security starts looking for a contractor to build the project with.
 * They find Photon Inc. from Chennai, India. [Website](http://www.photon.in/en/) (see sources)
 * The first versions are built and released, renamed from "Champion" to "Aviator".
 * Someone decides to call it "The Most Secure Browser Online" [Proof](https://www.google.ca/search?q=The+Most+Secure+Browser+Online)
 * In early 2014, the "Error138" is reported and disclosed first time. The one that is now an RCE. [Tweet](https://twitter.com/0x6D6172696F/statuses/444804907779244032)
 * Selfies with cool aviator shades appear online. [Tweet](https://twitter.com/RSnake/status/402836094435491841)
 * The level of Aviator's success however is low, not too many users download and actually use the tool.
 * Still , WhiteHat has to pay for the development. Which is indeed ongoing. [Proof](https://updates.aviatorbrowser.com/Win/releasenotes.txt)
 * Meanwhile, there's about 500 changes to the original Chromium code [Source](https://github.com/WhiteHatSecurity/Aviator/search?utf8=%E2%9C%93&q=whitehat)  
 * An audit or anything else direction reasonable security program never happens. Too expensive. No actual ROI. (well, see Aviator sources as a proof that no audit happened)
 * At some point the bills just become too much to pay, the money simply burns and nothing comes back. 
 * Meetings are held, people wondering what to do now.
 * An idea develops.  Aviator could be open sourced! That would mean no bills anymore and failure can be blamed on the community. Especially if bugs don't get fixed.
 * So, the community would be responsible for a branded WhiteHat Security product. And everyone wins. Well. Only WhiteHat. But still. (see comments on this ticket)
 * That however did not work out and a shit-storm happened. [Source](http://threatpost.com/google-engineers-critical-of-aviator-browser-security/110323)
 * Followed by a blog post blaming Google for being so rich. [Source](https://blog.whitehatsec.com/aviator-open-source-day-1/)

Other points about Aviator, the mediocre code, the lack of security bug tracker, no security response process, no technical documentation and the lack of anything else that is recommended if not even required to create and maintain a security critical piece of software is speaking for itself. 

I believe, that Aviator was meant to be a marketing tool that turned out to blow up and backfire on its creators. Aviator claims what it cannot deliver - "The Most Secure Browser Online". And now a blame game is being played to do reputation damage control.

So, to wrap it up, I don't see any altruist move here but a simple and failed attempt to do financial damage control and claim to have done something good while actually throwing off baggage and hiding the fact that Aviator is a dead horse.

I may be dead wrong with all these assumptions. Please challenge them or shout at me or do whatever you feel is adequate :)

**Comment by DinisCruz** [comment]
@cure53 interesting analysis. well I have to say that the first time aviator come out I asked Jeremiah and Robert to make it open source, since the sooner this process could be started the better (including the creation of a community). See [here](https://twitter.com/DinisCruz/status/392786861519024128)  and [here](https://twitter.com/DinisCruz/status/392804511619616768) from Oct 2013

The fact that Aviator was not originally open sourced did also raise in the back of my mind the question 'I bet they want to create a commercial/closed product' out of this. Which I always felt to be a VERY difficult proposition and going against the historical evolution of browsers.

I also think that the whole "The Most Secure Browser Online" was a massive mistake (which is clearly back firing now), specially since new code was being added to the chrome codebase which could introduce a number of other security vulnerabilities.

I hope that Aviator is not a dead horse since we do need more 'secure' browsers (special ones that come secure out of the box), but this might require a different approach. 

For starters somebody needs to:

* [Diff with the main @ChromiumDev distribution? What happens with chromium updates?](https://github.com/WhiteHatSecurity/Aviator/issues/27) 
* Write [Tests that represent the changes made to the original Chromium code](https://github.com/WhiteHatSecurity/Aviator/issues/37)
* Create [Automated builds in a CI environment](https://tm4tm.teammentor.net/teamMentor)
* Automate the [Upgrade to latest version of Chrome](https://github.com/WhiteHatSecurity/Aviator/issues/39)

As mentioned in [Upgrade to latest version of Chrome](https://github.com/WhiteHatSecurity/Aviator/issues/39) of I think that a key **design decision for the Aviator browser is to have version parity with Chrome releases, specially when there is a release (major or minor) that introduces security issues.**

**Comment by antisnatchor** [comment]
I completely agree with @cure53 analysis here, but I still want to add a few things:

 - I don't expect "security professionals/experts" to make statements like "the most secure browser". EVER. Full dot. There have been tons of examples of such statements that proved to be wrong, and this is yet another example. This honestly reminds me FireEye claims that they stop 0days...

 - I don't expect people believing that Aviator is "the most secure browser", while being:
   - just 95% Chromium code base (so we're not exactly in the situation to say: 'hey, thanks WhiteHat!'. I would say instead: 'hey, thanks Google')
   - plus a bunch of extensions
   - plus an additional 5% code base written by some C++ contractor that likes to concatenate strings and passing them straight-away to a dangerous function. 

I mean, seriously, do you trust more the smart guys of the Google Security Team (with skills, money, resources, time) or WhiteHat (with probably with less skills, money, resources, time)??

 - I don't expect people really need the ability to run OS applications from the browser URL bar, do you??? Honestly, what was the reason behind that regardless of the ridiculous bug??

Another thing..Lets say I have a 1day UAF+Sandbox bypass, already patched in latest Chrome, but thanks to the fact latest Aviator hasn't the latest Chromium patches applied (as its code is behind X releases), I can still use it against Aviator. Does this makes it "the most secure browser"?

Moreover, think about the following: you get the Firefox source, add a bunch of extensions like NoScript/AdBlocker/Whatever, and add a 5% of your own code, so you now have the right to call it "the most secure browser". How nice is that? 

Last thing to add is the following: the "proliferation" of browsers isn't something that help having a more secure browser. See Avant/Maxthon browsers and the ridiculous bugs that @malerisch found on them, or for instance the situation of Safari...

**Comment by continuumsecurity** [comment]
There's still the unanswered question posed in Google's comments: why isn't Aviator a chrome extension rather than a fork?

**Comment by DinisCruz** [comment]
@continuumsecurity I wrote an chrome extension (see [chrome-repl](https://github.com/o2platform/chrome-repl) ) and I'm not sure that the designer changes to native chrome behaviour can be from there

**Comment by dgc-wh** [comment]
This is addressed in #45

**Comment by DinisCruz** [comment]
Great, is there a test for it?
